File: Python/Sliding Window/Best Time to Buy and Sell Stock.py
class Solution:
    def maxProfit(self, prices: List[int]) -> int:

        ans = 0
        minNum = sys.maxsize

        for price in prices:
            minNum = min(minNum,price)
            ans = max(ans, price - minNum)

        return ans

#For most practical purposes, sys.maxsize serves as a suitable substitute for INT_MAX in Python.

#Kadane's Algorithm
#the logic is to calculate the difference (maxCur += prices[i] - prices[i-1]) of the original array, and find a contiguous subarray giving maximum profit. If the difference falls below 0, reset it to zero.

File: C++/Sliding Window/Best Time to Buy and Sell Stock.cpp
class Solution {
public:
    int maxProfit(vector<int>& prices) {
        // int n = prices.size(), ans = 0 , max_elem_till_now=prices[n-1];

        // for( int i = n-2 ; i>=0 ; i--){
        //     if(prices[i]<max_elem_till_now)
        //     ans = max(ans, abs(prices[i]-max_elem_till_now));
        //    // min_elem_till_now = min(min_elem_till_now, prices[i]);
        //     max_elem_till_now = max (max_elem_till_now, prices[i]);

        // }
        // return ans;


        int n = prices.size();
        if(n<=1)
            return 0;
        int ans = 0, max_num_till_now= prices[n-1];

        for( int i=n-2 ; i>=0 ; i--){
            if(prices[i] < max_num_till_now)
                ans = max(ans,(max_num_till_now - prices[i]));

            max_num_till_now = max(max_num_till_now,prices[i]);
        }
        return ans;

    }
};